File: client/src/utils/passwordPolicy.js
export const PASSWORD_POLICY_MESSAGE =
    "A senha deve ter no mínimo 9 caracteres, incluindo pelo menos 1 letra, 1 número e 1 caractere especial."

export const getPasswordRequirementChecks = (password, confirmPassword = "") => {
    const value = String(password || "")
    const confirmation = String(confirmPassword || "")

    return [
        {
            id: "length",
            label: "ao menos 9 caracteres",
            met: value.length >= 9,
        },
        {
            id: "letter",
            label: "ao menos uma letra",
            met: /[A-Za-z]/.test(value),
        },
        {
            id: "number",
            label: "ao menos 1 numero",
            met: /\d/.test(value),
        },
        {
            id: "special",
            label: "ao menos um caracter especial (!@#$%*,./<>)",
            met: /[^A-Za-z0-9]/.test(value),
        },
        {
            id: "confirmation",
            label: "a confirmacao da senha deve coincidir com a senha",
            met: value.length > 0 && confirmation.length > 0 && value === confirmation,
        },
    ]
}

export const isStrongPassword = (password) =>
    getPasswordRequirementChecks(password, password).slice(0, 4).every((item) => item.met)

export const getPasswordValidationMessage = (password) =>
    isStrongPassword(password) ? "" : PASSWORD_POLICY_MESSAGE

File: mobile/RendaTop.App/Services/PasswordPolicy.cs
namespace RendaTop.App.Services;

public static class PasswordPolicy
{
    public const string Message =
        "A senha deve ter no minimo 9 caracteres, incluindo pelo menos 1 letra, 1 numero e 1 caractere especial.";

    public static string Validate(string password, string confirmation)
    {
        if (password != confirmation)
            return "As senhas nao conferem.";

        if (password.Length < 9 ||
            !password.Any(char.IsLetter) ||
            !password.Any(char.IsDigit) ||
            password.All(char.IsLetterOrDigit))
        {
            return Message;
        }

        return string.Empty;
    }
}
